File: kotlin/2015/day01b/Program.kt
import java.io.File

fun usage() {
    System.err.println("usage: java -jar Program.jar <input file>")
    kotlin.system.exitProcess(1)
}

fun process(content: String) : Int {
    var floors = 0
    var pos = 0

    for ( ch in content ) {
        pos += 1;
        when ( ch ) {
            '(' -> floors += 1
            ')' -> floors -= 1
            else -> {}
        }
        if ( floors < 0 ) {
            return pos
        }
    }

    return 0
}

fun main(args: Array<String>) {
    if ( args.isEmpty()) {
        usage()
    }
    val filename = args[0]
    try {
        val content = File(filename).readText()
        val result = process(content)
        println("result = $result")
    } catch ( e: Exception ) {
        System.err.println("error reading file: ${e.message}")
    }
}

File: scala/2015/day01b/Program.scala
import scala.io.Source
import scala.util.Using
import scala.sys.exit

object Program {
  def usage(): Unit = {
    Console.err.println("usage: scala Program <input file>")
    sys.exit(1)
  }

  def process(filename: String): Int = {
    val content = Using.resource(Source.fromFile(filename))(_.mkString)
    def checkPosition(chars: List[Char], floor: Int, pos: Int): Int = (chars, floor, pos) match {
      case (_, floor, pos) if floor < 0 => pos
      case ('('::xs, floor, pos) => checkPosition(xs, floor + 1, pos + 1)
      case (')'::xs, floor, pos) => checkPosition(xs, floor - 1, pos + 1)
      case (_::xs, floor, pos) => checkPosition(xs, floor, pos)
      case _ => 0
    }
    checkPosition(content.toList, 0, 0)
  }

  def main(args: Array[String]): Unit = {
    args.toList match {
      case filename :: Nil =>
        val result = process(filename)
        println(s"result = $result")
      case _ =>
        usage()
    }
  }
}
